File: algorithm/python/SecondMinimumNodeInABinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findSecondMinimumValue(self, root):

        ans, rootvalue = -1, root.val

        def dfs(node):
            nonlocal ans
            if not node:
                return
            if ans != -1 and node.val >= ans:
                return
            if node.val > rootvalue:
                ans = node.val
            
            dfs(node.left)
            dfs(node.right)

        dfs(root)
        return ans

File: algorithm/JavaScript/SecondMinimumNodeInABinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
 var findSecondMinimumValue = function(root) {
  let res = -1
  const rootValue = root.val

  const dfs = (node) => {
    if (node === null) {
      return
    }
    if (res !== -1 && node.val >= res) {
      return
    }
    if (node.val > rootValue) {
      res = node.val
    }
    dfs(node.left)
    dfs(node.right)
  }

  dfs(root)
  return res
};